File: csharp/anagram/Anagram.cs
﻿using System;
using System.Collections.Generic;

public class Anagram
{
	// Alphabetically sorted version of original phrase for letter set matching
	private string _key;
	// Original phrase for checking self identity of anagram candidate
	private string _phrase;

	// Constructor stores alphabetized key and original phrase
	public Anagram (string phrase) {
		_key = CreateKey (phrase);
		_phrase = phrase.ToLower();
	}

	// Of a set of input words,
	// returns the set that are matching anagrams of original phrase
	public string[] Match (string[] words) {
		var matches = new List<string> ();
		foreach (var word in words) {
			if (IsAnagram(word.ToLower())) {
				matches.Add(word);
			}
		}
		return matches.ToArray();
	}

	// Checks the candidate phrase against the stored key and original phrase
	// Assumes candidate phrase is all lower case
	private bool IsAnagram (string candidate) {
		var ret = true;
		if (_phrase == candidate) {
			// A word cannot be an anagram of its identical self
			ret = false;
		}
		if (_key != CreateKey (candidate)) {
			// Set of letters don't match for anagram
			ret = false;
		}
		return ret;
	}

	// CreateKey returns the alphabetized order of the characters
	// for comparisons between potential anagrams
	private static string CreateKey (string phrase) {
		// Assuming ASCII characters (no extended Unicode)
		var arrayToSort = phrase.ToLower().ToCharArray ();
		Array.Sort (arrayToSort);
		return new string (arrayToSort);
	}
}

File: python/anagram/anagram.py
def detect_anagrams(target, candidates):
    """ Given a target string and a space-separated list of candidates,
        find the anagrams of the target string """
    matches = []
    for candidate in candidates:
        if is_anagram(target, candidate):
            matches.append(candidate)
    return matches

def is_anagram(target, candidate):
    """ Returns true if the candidate string is an anagram of the target """
    if target == candidate:
        # Identical strings are not anagrams of each other
        return False
    elif get_key(target) == get_key(candidate):
        # The number of letters match, so the two words are anagrams
        return True
    else:
        # Letters on either side don't match up as anagrams
        return False

def get_key(text):
    """ Generates a key from a string,
        which is the alphabetically sorted letters """
    return sorted(text)
